### DISCUSSION TITLE
How do I style gt tables in pdf documents

### DISCUSSION THREAD

**Discussion by olivroy** [quarto-dev/quarto-cli#9408] [Q&A]
### Description

I would like to enable styling for a gt table in pdf output. But I can't find any resources on how to make row striping or adding horizontal lines working in pdf

````qmd

---
title: "gt table styling"
format: 
  pdf:
   include-in-header: 
    text: |
      \usepackage[margin=0.1in]{geometry}
      % from ?gt::as_latex
      \usepackage{booktabs, caption, longtable, colortbl, array}
  html: default
editor: visual
# followed advice from https://gt.rstudio.com/reference/as_latex.html 
execute: 
  echo: false
---

## A LaTeX table

```{r}
library(gt)
my_gt <- 
exibble |> 
  gt::gt() |> 
    gt::tab_style(
    style = gt::cell_borders(sides = "top", color = "gray48"),
    locations = gt::cells_body()
  ) |> 
  gt::opt_row_striping(row_striping = TRUE)
```

::: {.content-visible when-format="pdf"}

With `gt::as_latex()`

```{r}
#| output: asis
my_gt |> 
  gt::as_latex()
```

:::

regular quarto rendering

```{r}
my_gt
```

````

With Quarto 1.4.553 and gt 0.10.1 on R 4.3.3 on Windows + using tinytex if that is useful

However, both results show no styling.
![image](https://github.com/quarto-dev/quarto-cli/assets/52606734/5c1e82fa-1288-459f-ae25-2cca54c74b53)

**Comment by mcanouil** [comment]
That's not a Quarto question and it is a bit out of Quarto support scope.
Be sure to read the documentation of the package you are using.
You can check what you are outputting using `keep-*` keys.

For instance, the LaTeX you are emitting is (output obtained directly from the R code in R):

```latex
\begin{longtable}{rlcrrrrll}
\toprule
num & char & fctr & date & time & datetime & currency & row & group \\ 
\midrule\addlinespace[2.5pt]
1.111e-01 & apricot & one & 2015-01-15 & 13:35 & 2018-01-01 02:22 & 49.950 & row\_1 & grp\_a \\ 
2.222e+00 & banana & two & 2015-02-15 & 14:40 & 2018-02-02 14:33 & 17.950 & row\_2 & grp\_a \\ 
3.333e+01 & coconut & three & 2015-03-15 & 15:45 & 2018-03-03 03:44 & 1.390 & row\_3 & grp\_a \\ 
4.444e+02 & durian & four & 2015-04-15 & 16:50 & 2018-04-04 15:55 & 65100.000 & row\_4 & grp\_a \\ 
5.550e+03 & NA & five & 2015-05-15 & 17:55 & 2018-05-05 04:00 & 1325.810 & row\_5 & grp\_b \\ 
NA & fig & six & 2015-06-15 & NA & 2018-06-06 16:11 & 13.255 & row\_6 & grp\_b \\ 
7.770e+05 & grapefruit & seven & NA & 19:10 & 2018-07-07 05:22 & NA & row\_7 & grp\_b \\ 
8.880e+06 & honeydew & eight & 2015-08-15 & 20:20 & NA & 0.440 & row\_8 & grp\_b \\ 
\bottomrule
\end{longtable}
```

**Reply by olivroy** [reply]
Ok agreed this is not a Quarto issue.

However, I think that it would be relevant to add a section on Tables in the pdf section

https://quarto.org/docs/output-formats/pdf-basics.html

I didn't find anything, which suggests that Quarto can only render tables like this to pdf:

![image](https://github.com/quarto-dev/quarto-cli/assets/52606734/6f8f55e4-c8da-412f-b2e3-4a91d76d2c3f)

It is difficult to figure out what can be solved with Quarto, LaTeX or gt. (disabling processing is also a confusion here)

Turns out the problem was the gt doesn't respect its `opt_row_striping(row_striping = TRUE)` when format: pdf. So I opened an issue there https://github.com/rstudio/gt/issues/1625.

In the mean time, I found a working workaround, but it took me a while to figure it out.

````qmd
---
title: title
format: pdf
---

```{r}
library(gt)
# compute manually even rows
rows_to_span <- which(seq_len(nrow(exibble)) %% 2 == 0)

exibble |> 
  gt() |>
  gt::tab_style(
    style = gt::cell_fill("gray85"), # cell_fill() seems to have been implemented.
    locations = gt::cells_body(rows = rows_to_span)
  ) |>
 gt::tab_options(quarto.disable_processing = TRUE)

```
````


However, I think it would be great if Quarto had a nicer default look for tables in pdf outputs, is this something that is planned?

The default Quarto processing for html on the other hand has all I was looking to add to my pdf table:

* striping
* horizontal borders

I couldn't find a clear way to do this with Quarto, either via gt styling or via adding stuff as LaTeX preamble that would override table styling.

Because to get the colors, we need to disable processing, we may lose cross-refs.

**Reply by mcanouil** [reply]
> However, I think that it would be relevant to add a section on Tables in the pdf section
> https://quarto.org/docs/output-formats/pdf-basics.html
> I didn't find anything, which suggests that Quarto can only render tables like this to pdf:

Quarto can use whatever you give it, there are no restrictions there, so nothing to add in the documentation.
Here `gt` provide a LaTeX code that does not have row striping.
Quarto cannot render what does not exist.
Quarto is not responsible of what you use to emit code.

> In the mean time, I found a working workaround, but it took me a while to figure it out.

This is described in both Quarto and `gt` documentation.
Note that here, you are outputting HTML table, which Quarto can process to the AST format then back to LaTeX.
This is why you can use the HTML processing done by Quarto.
See:
- https://quarto.org/docs/authoring/tables.html#html-tables
- https://quarto.org/docs/authoring/tables.html#disabling-quarto-table-processing

> However, I think it would be great if Quarto had a nicer default look for tables in pdf outputs, is this something that is planned?

Quarto does not really set theming. You are free to change or use anything.
"nicer default look" is a very subjective comment that not all users will agree upon.
Note that the templates are mostly coming straight from Pandoc.
Create your own PDF format (eventually package it as a Quarto extension) and use it. That's what the extension mechanism is for.
Unless the aesthetics has accessibility issues, changes are very unlikely to occur in Quarto (or Pandoc for that matter).

> Because to get the colors, we need to disable processing, we may lose cross-refs.

Cross-reference should still work. The option disable "table processing" not all processing.

**Reply by olivroy** [reply]
Ok. Thanks for taking the time for the clarifications!
